src/pjpersist/datamanager.py

```diff
@@ -665,6 +667,15 @@ def _join_txn(self):
             self._needs_to_join = False
             self._begin(transaction)
 
+    def _doom_txn(self):
+        try:
+            self.transaction_manager.doom()
+            return True
+        except ValueError:
+            # We are not in active trasaction phase,
+            # so it cannot be doomed.
+            return False
```

Comment:
Not sure ignoring exception here is a right thing to do. Why would we doom the transaction when we are not in "transaction phase"?

Reply:
auto create table feature during a commit.  flush too.  didn't find a nicer way to figure out if we're inside of a commit or not.

Reply:
Isn't flush is part of transaction? Errors during flush should doom the transaction, FWICT.  Auto-create should be part of transaction too, as I understand.

Reply:
Hmm, maybe I misunderstand something.
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/zope/tests/test_container.py", line 1947, in test_conflict_SimplePJContainer_key_2
    transaction.commit()
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_manager.py", line 257, in commit
    return self.manager.commit()
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_manager.py", line 134, in commit
    return self.get().commit()
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_transaction.py", line 282, in commit
    reraise(t, v, tb)
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_compat.py", line 49, in reraise
    raise value
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_transaction.py", line 273, in commit
    self._commitResources()
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_transaction.py", line 456, in _commitResources
    reraise(t, v, tb)
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_compat.py", line 49, in reraise
    raise value
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_transaction.py", line 430, in _commitResources
    rm.commit(self)
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/datamanager.py", line 879, in commit
    self.flush()
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/datamanager.py", line 632, in flush
    self._writer.store(docobj)
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/serialize.py", line 445, in store
    doc = self.get_state(obj.__getstate__(), obj)
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/serialize.py", line 371, in get_state
    data.append((key, self.get_state(value, pobj)))
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/serialize.py", line 371, in get_state
    data.append((key, self.get_state(value, pobj)))
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/serialize.py", line 390, in get_state
    return self.get_persistent_state(obj)
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/serialize.py", line 303, in get_persistent_state
    dbref = self.store(obj, ref_only=True)
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/serialize.py", line 457, in store
    doc_id = self._jar._insert_doc(
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/datamanager.py", line 553, in _insert_doc
    cur.execute(sql, tuple(values),
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/datamanager.py", line 251, in execute
    self._autoCreateTables(sql, args, beacon)
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/datamanager.py", line 214, in _autoCreateTables
    self.datamanager._doom_txn()
  File "/home/justas/src/shoobx/pjpersist/src/pjpersist/datamanager.py", line 673, in _doom_txn
    transaction.doom()
  File "/home/justas/src/shoobx/pjpersist/ve/lib/python3.8/site-packages/transaction/_transaction.py", line 171, in doom
    raise ValueError('non-doomable')
ValueError: non-doomable
```

Reply:
Does anything suggests the exception needs to be ignored?

Reply:
Yes, flush being called from `_commitResources()`.  

PJDataManager is a resource, it is not a transaction or a transaction manager.  Resources are committed outside of active phase, in the committing phase.   Python transactions do not allow being doomed, while being committed.

See https://github.com/zopefoundation/transaction/blob/master/src/transaction/_transaction.py#L273

Reply:
The `ValueError` that is raised from `doom()` seem to be intended as a safeguard. It doesn't look like it is expected to be intercepted. The code that ignores the safeguard always looks suspicious and the comment doesn't help much understand why it is happening, just translates the `doom()` source code to English.

I'd probably specifically avoid dooming transaction in COMMITTING state here, instead of catching the exception.

Functionally that would be equivalent, so I'm fine with this code as it is (since I personally understand the motivation behind it now).